### PR TITLE
[MRv2] FlashAttention PCP support for GQA on MRv2

### PR DESCRIPTION
**Pull request by JaredforReal**
# Hy3-290B-FP8 PCP/DCP Benchmark in 8*H100

Model: `tencent/Hy3-FP8` (HYV3ForCausalLM, GQA 64Q/8KV, 192-expert MoE + 1 shared,
80 layers, hidden 4096, 279 GB weights). Hardware: 8×H100 80GB.

All configs use `--enforce-eager`, `--enable-expert-parallel`,
`--no-enable-flashinfer-autotune` (skips the FlashInfer warmup that OOMs under
TP1 where all 64 attention heads land on one GPU), `--gpu-memory-utilization 0.82`
(headroom for warmup activations; 0.88 OOMs on TP2+PCP4), `--max-model-len 32800`
(TP1+PCP8 has only 12.4 GiB KV cache; 40960 needs 12.5 GiB), `--safetensors-load-strategy
prefetch` (boot 9 min → 2.5 min), `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True`.

## Config matrix — all 7 use exactly 8 GPUs

`TP × PCP = 8` throughout; `DCP == PCP` so DCP ranks share PCP ranks (no extra
world size). EP spans TP × PCP = 8 in every config.

| # | Config | TP | PCP | DCP | KV cache |
|---|---|---|---|---|---|
| 1 | tp8 | 8 | 1 | 1 | replicated (baseline) |
| 2 | tp4_pcp2 | 4 | 2 | 1 | replicated |
| 3 | tp2_pcp4 | 2 | 4 | 1 | replicated |
| 4 | tp1_pcp8 | 1 | 8 | 1 | replicated (pure PCP) |
| 5 | tp4_pcp2_dcp2_sharded | 4 | 2 | 2 | sharded (1/dcp) |
| 6 | tp2_pcp4_dcp4_sharded | 2 | 4 | 4 | sharded (1/dcp) |
| 7 | tp1_pcp8_dcp8_sharded | 1 | 8 | 8 | sharded (1/dcp) |

## TPOT — decode latency (mean / P99 ms)

ShareGPT subset, prefix caching **on** (decode is prefix-insensitive; matches
how the Qwen-30B baselines were taken). Concurrency 1 / 16 / 32.

| Config | c=1 | c=16 | c=32 | c1→32 |
|---|---|---|---|---|
| TP8 (baseline) | 158 / 161 | 158 / 165 | 162 / 165 | +2.5% |
| TP4+PCP2 | 177 / 181 | 180 / 183 | 180 / 183 | +1.7% |
| TP2+PCP4 | 174 / 180 | 177 / 180 | 178 / 182 | +2.3% |
| TP1+PCP8 | 167 / 171 | 170 / 175 | 174 / 179 | +4.2% |
| TP4+PCP2+DCP2 | 205 / 215 | 213 / 225 | 213 / 219 | +3.9% |
| TP2+PCP4+DCP4 | 205 / 208 | 215 / 220 | 215 / 220 | +4.9% |
| TP1+PCP8+DCP8 | 195 / 198 | 202 / 208 | 206 / 215 | +5.6% |

## TTFT — prefill latency (median ms)

Random-input probe, prefix caching **off** + 6 warmup requests discarded,
20 measured.

| Config | 4K | 16K | 32K |
|---|---|---|---|
| TP8 (baseline) | 256 | 926 | 1973 |
| TP4+PCP2 | 243 | 875 | 1858 |
| TP2+PCP4 | 238 | 832 | 1741 |
| TP1+PCP8 | 226 | 783 | 1682 |
| TP4+PCP2+DCP2 | 244 | 911 | 1979 |
| TP2+PCP4+DCP4 | 233 | 891 | 1980 |
| TP1+PCP8+DCP8 | 228 | 927 | 2123 |

## Accuracy & KV capacity

GSM8K 200 questions, 5-shot, greedy, conc 256.

| Config | GSM8K | KV tokens | DCP capacity win |
|---|---|---|---|
| TP8 | 0.885 | 735,056 | 1× |
| TP4+PCP2 | 0.905 | 338,288 | — |
| TP2+PCP4 | 0.915 | 139,840 | — |
| TP1+PCP8 | 0.940 | 40,736 | — |
| TP4+PCP2+DCP2 | 0.915 | 676,576 | **2.0×** vs TP4+PCP2 |
| TP2+PCP4+DCP4 | 0.895 | 558,814 | **4.0×** vs TP2+PCP4 |
| TP1+PCP8+DCP8 | 0.910 | 324,936 | **8.0×** vs TP1+PCP8 |

GSM8K spread (0.885–0.940) is 200-question sample noise (each question = 0.5%,
so ±2–3 questions = ±1–1.5%), not a regression.

---

# Qwen3-30B-A3B-FP8 PCP/DCP Benchmark in 4xH100

Model: `Qwen3-30B-A3B-FP8` (GQA, MoE, 48 layers). Hardware: 4×H100 80GB.

All configs use `--enforce-eager`, `--enable-expert-parallel`,
`--gpu-memory-utilization 0.88`, `--max-model-len 40960`. 

## Config matrix — all 5 use 4 GPUs

When `dcp == pcp` the DCP ranks share the PCP ranks, so the CP world size is
just `pcp`. Config 4 (PCP2+DCP2) therefore runs on **2 GPUs**, not 4.

| # | Config | TP | PCP | DCP | KV cache | GPUs |
|---|---|---|---|---|---|---|
| 1 | tp4 | 4 | 1 | 1 | replicated (baseline) | 4 |
| 2 | tp2_pcp2 | 2 | 2 | 1 | replicated | 4 |
| 3 | tp1_pcp4 | 1 | 4 | 1 | replicated (pure PCP) | 4 |
| 4 | tp1_pcp2_dcp2_sharded | 1 | 2 | 2 | sharded (1/dcp) | **2** |
| 5 | tp1_pcp4_dcp4_sharded | 1 | 4 | 4 | sharded (1/dcp) | 4 |

## TPOT — decode latency (mean / P99 ms)

ShareGPT subset: concurrency 1 / 16 / 32.

| Config | c=1 | c=16 | c=32 | c1→32 |
|---|---|---|---|---|
| TP4 (baseline) | 60.2 / 64.5 | 60.9 / 64.1 | 65.7 / 76.5 | +9.1% |
| TP2+PCP2 | 73.6 / 78.9 | 74.7 / 79.1 | 79.5 / 84.9 | +8.0% |
| TP1+PCP4 | 68.6 / 70.2 | 70.2 / 76.1 | 75.9 / 80.6 | +10.6% |
| TP1+PCP2+DCP2 | 83.9 / 87.7 | 89.6 / 95.7 | 94.6 / 100.8 | +12.8% |
| TP1+PCP4+DCP4 | 85.6 / 91.7 | 90.3 / 95.1 | 96.4 / 105.3 | +12.6% |

The PCP gain is from decode steps no longer doing a per-layer `forward_context`
lookup + store, nor building the pcp-wide gathered slot mapping. DCP configs
gain less because their per-step LSE-combine collective dominates what was
removed.

## TTFT — prefill latency (median ms)

Random-input probe, prefix caching **off** + 6 warmup requests discarded,
20 measured.

| Config | 4K | 16K | 32K |
|---|---|---|---|
| TP4 (baseline) | 96 | 289 | 645 |
| TP2+PCP2 | 116 | 301 | 648 |
| TP1+PCP4 | 110 | 287 | 622 |
| TP1+PCP2+DCP2 | 109 | 423 | 1038 |
| TP1+PCP4+DCP4 | 114 | 333 | 771 |

The three no-DCP configs converge at 32K (622–648 ms — same parallelised
prefill). DCP4 is +126 ms (+20%) over that floor; DCP2 is much worse but runs
on **2 GPUs**.

## Accuracy & KV capacity

GSM8K 1319 questions, 5-shot, greedy, conc 256.

| Config | GSM8K | KV tokens | Max concurrency | DCP capacity win |
|---|---|---|---|---|
| TP4 | 0.901 | 2,657,504 | — | — |
| TP2+PCP2 | 0.905 | 1,302,384 | — | — |
| TP1+PCP4 | 0.886 | 626,016 | — | — |
| TP1+PCP2+DCP2 | 0.896 | 1,132,096 | — | **1.81×** vs TP1+PCP4 |
| TP1+PCP4+DCP4 | 0.891 | 2,504,064 | 61.13× | **4.00×** vs TP1+PCP4 |

### CODE REVIEW

File: vllm/v1/attention/backends/flash_attn.py [hunk 1]
```diff
@@ -1131,7 +1144,7 @@ def do_kv_cache_update(
             layer._v_scale,
         )
 
-    def _forward_with_dcp(
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn.py:1134]: Make sure to update FlashAttentionDiffKV too please

File: vllm/v1/attention/backends/flash_attn.py [hunk 2]
```diff
@@ -1140,172 +1153,146 @@ def _forward_with_dcp(
...
+            softcap=self.logits_soft_cap,
+            return_softmax_lse=True,
+            fa_version=self.vllm_flash_attn_version,
+            num_splits=attn_metadata.max_num_splits,
         )
-        dcp_context_out = dcp_context_out_workspace[:n]
 
-        if split_dcp_context:
-            # TODO: Remove this DCP + FA2 mixed decode/prefill workaround once
-            # FA4 supports this Qwen3.5 shape.
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn.py:1215]: Are you sure we can remove this?

File: vllm/v1/attention/backends/flash_attn.py [hunk 3]
```diff
@@ -945,14 +942,15 @@ def forward(
             v_descale = layer._v_scale.expand(descale_shape)
 
             if self.dcp_world_size > 1:
-                self._forward_with_dcp(
-                    query[:num_actual_tokens],
+                self._forward_dcp(
+                    query,
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn.py:946]: Why get rid of this slice? Are we computing over padding rows now?

File: vllm/v1/attention/backends/flash_attn.py [hunk 4]
```diff
@@ -1140,172 +1153,146 @@ def _forward_with_dcp(
...
+            cu_q = attn_metadata.query_start_loc
+            max_q = attn_metadata.max_query_len
+            ctx_lens = attn_metadata.dcp_context_kv_lens
+            max_ctx = attn_metadata.max_dcp_context_kv_len
+            ctx_cu = None  # not needed — pure DCP uses seqused_k=0
+            block_table = attn_metadata.block_table
+            fa_kwargs["scheduler_metadata"] = attn_metadata.scheduler_metadata
+
+        descale_shape = (cu_q.shape[0] - 1, self.num_kv_heads)
+        ctx_out, ctx_lse = flash_attn_varlen_func(
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn.py:1261]: Please use the workspace and pass `out` so we aren't allocating each step

**Review by MatthewBonanni — COMMENTED** [review]
Thanks! A few comments
